File: mobile/lib/services/user_mutations.dart
import 'package:supabase_flutter/supabase_flutter.dart';

import '../models/user_profile.dart';

class UserMutationResult {
  const UserMutationResult({required this.ok, this.error, this.mode});

  final bool ok;
  final String? error;
  final String? mode;
}

bool isProfileAccessible(UserProfile? profile) {
  if (profile == null) return false;
  if (profile.deletedAt != null) return false;
  if (!profile.actif) return false;
  return true;
}

bool canWriteAdmin(UserProfile? profile) =>
    profile?.role == 'admin' || profile?.role == 'superadmin';

/// Mutations utilisateurs — profiles uniquement (aligné web).
class UserMutations {
  UserMutations(this._client);

  final SupabaseClient _client;

  UserMutationResult _denied() =>
      const UserMutationResult(ok: false, error: 'Accès admin requis');

  Future<UserMutationResult> updateUserRole({
    required String userId,
    required String newRole,
    required UserProfile? actorProfile,
  }) async {
    if (!canWriteAdmin(actorProfile)) return _denied();
    try {
      await _client.rpc('fn_admin_update_user_role', params: {
        'p_user_id': userId,
        'p_new_role': newRole,
      });
      return UserMutationResult(ok: true, mode: 'role_updated');
    } on PostgrestException catch (e) {
      return UserMutationResult(ok: false, error: e.message);
    }
  }

  Future<UserMutationResult> softBanUser({
    required String userId,
    required UserProfile? actorProfile,
    String? reason,
  }) async {
    if (!canWriteAdmin(actorProfile)) return _denied();
    try {
      await _client.rpc('fn_admin_soft_ban_user', params: {
        'p_user_id': userId,
        'p_reason': reason,
      });
      return const UserMutationResult(ok: true, mode: 'soft_ban');
    } on PostgrestException catch (e) {
      return UserMutationResult(ok: false, error: e.message);
    }
  }

  Future<UserMutationResult> reactivateUser({
    required String userId,
    required UserProfile? actorProfile,
  }) async {
    if (!canWriteAdmin(actorProfile)) return _denied();
    try {
      await _client.rpc('fn_admin_reactivate_user', params: {
        'p_user_id': userId,
      });
      return const UserMutationResult(ok: true, mode: 'reactivated');
    } on PostgrestException catch (e) {
      return UserMutationResult(ok: false, error: e.message);
    }
  }

  Future<UserMutationResult> hardDeleteUser({
    required String userId,
    required UserProfile? actorProfile,
  }) async {
    if (!canWriteAdmin(actorProfile)) return _denied();
    try {
      await _client.rpc('fn_admin_hard_delete_user', params: {
        'p_user_id': userId,
      });
      return const UserMutationResult(ok: true, mode: 'hard_delete');
    } on PostgrestException catch (e) {
      return UserMutationResult(ok: false, error: e.message);
    }
  }

  Future<UserMutationResult> toggleUserActive({
    required String userId,
    required bool currentlyActive,
    required UserProfile? actorProfile,
  }) {
    if (currentlyActive) {
      return softBanUser(userId: userId, actorProfile: actorProfile);
    }
    return reactivateUser(userId: userId, actorProfile: actorProfile);
  }

  Future<UserMutationResult> toggleUserVerified({
    required String userId,
    required bool currentlyVerified,
    required UserProfile? actorProfile,
  }) async {
    if (!canWriteAdmin(actorProfile)) return _denied();
    try {
      await _client
          .from('profiles')
          .update({
            'verifie': !currentlyVerified,
            'updated_at': DateTime.now().toUtc().toIso8601String(),
          })
          .eq('id', userId)
          .filter('deleted_at', 'is', null);
      return UserMutationResult(ok: true, mode: 'verified_toggled');
    } on PostgrestException catch (e) {
      return UserMutationResult(ok: false, error: e.message);
    }
  }
}

File: src/lib/userMutations.js
import { supabase } from "./supabase";
import { canWriteAdmin } from "./roles";

function normalizeError(error) {
  if (!error) return null;
  return error.message || String(error);
}

function assertAdminWrite(profile) {
  if (!canWriteAdmin(profile)) {
    return { ok: false, error: "Accès refusé — droits admin requis" };
  }
  return null;
}

/**
 * Change le rôle d'un utilisateur (profiles uniquement).
 */
export async function updateUserRole({ userId, newRole, actorProfile }) {
  const denied = assertAdminWrite(actorProfile);
  if (denied) return denied;

  const { data, error } = await supabase.rpc("fn_admin_update_user_role", {
    p_user_id: userId,
    p_new_role: newRole,
  });

  if (error) return { ok: false, error: normalizeError(error) };
  return { ok: true, role: data?.role || newRole };
}

/**
 * Suspendre un utilisateur (soft ban, réversible).
 */
export async function softBanUser({ userId, reason, actorProfile }) {
  const denied = assertAdminWrite(actorProfile);
  if (denied) return denied;

  const { data, error } = await supabase.rpc("fn_admin_soft_ban_user", {
    p_user_id: userId,
    p_reason: reason || null,
  });

  if (error) return { ok: false, error: normalizeError(error) };
  return { ok: true, mode: data?.mode || "soft_ban" };
}

/**
 * Réactiver un utilisateur suspendu.
 */
export async function reactivateUser({ userId, actorProfile }) {
  const denied = assertAdminWrite(actorProfile);
  if (denied) return denied;

  const { data, error } = await supabase.rpc("fn_admin_reactivate_user", {
    p_user_id: userId,
  });

  if (error) return { ok: false, error: normalizeError(error) };
  return { ok: true, mode: data?.mode || "reactivated" };
}

/**
 * Suppression définitive (anonymisation du profil).
 */
export async function hardDeleteUser({ userId, actorProfile }) {
  const denied = assertAdminWrite(actorProfile);
  if (denied) return denied;

  const { data, error } = await supabase.rpc("fn_admin_hard_delete_user", {
    p_user_id: userId,
  });

  if (error) return { ok: false, error: normalizeError(error) };
  return { ok: true, mode: data?.mode || "hard_delete" };
}

/**
 * Bascule actif / suspendu (soft ban ou réactivation).
 */
export async function toggleUserActive({ userId, currentlyActive, actorProfile }) {
  if (currentlyActive !== false) {
    return softBanUser({ userId, actorProfile });
  }
  return reactivateUser({ userId, actorProfile });
}

/**
 * Bascule vérification utilisateur (profiles).
 */
export async function toggleUserVerified({ userId, currentlyVerified, actorProfile }) {
  const denied = assertAdminWrite(actorProfile);
  if (denied) return denied;

  const { error } = await supabase
    .from("profiles")
    .update({ verifie: !currentlyVerified, updated_at: new Date().toISOString() })
    .eq("id", userId)
    .is("deleted_at", null);

  if (error) return { ok: false, error: normalizeError(error) };
  return { ok: true, verifie: !currentlyVerified };
}

/** Profil actif et non supprimé */
export function isProfileAccessible(profile) {
  if (!profile) return false;
  if (profile.deleted_at) return false;
  if (profile.actif === false) return false;
  return true;
}
